lib/roken/issuid.c

```diff
@@ -118,98 +179,220 @@ issuid(void)
-        return 0;
-#endif
+    errno = save_errno;
 
-    /* rk_getauxval() does set errno */
-    if (errno == 0)
-        return 0;
     /*
-     * Fall through if we have getauxval() but we didn't have (or don't
-     * know if we don't have) the aux entries that we needed.
+     * This pre-processor condition could be all &&s, but that could
+     * cause a warning that seen is set but never used.
+     *
+     * In practice if any one of these four macros is defined then all
+     * of them will be.
      */
-#endif /* USE_RK_GETAUXVAL */
+#if (defined(AT_EUID) && defined(AT_UID)) || (defined(AT_EGID) && defined(AT_GID))
+    if (seen == 15) {
+        errno = save_errno;
+        return we_are_suid = 0;
+    }
+#endif
 
 #if defined(HAVE_ISSETUGID)
     /*
-     * If we have issetugid(), use it.
+     * If we have issetugid(), use it.  Illumos' and OpenBSD's
+     * issetugid() works correctly.
      *
-     * We may lose on some BSDs.  This manifests as, for example,
-     * gss_store_cred() not honoring KRB5CCNAME.
+     * On NetBSD and FreeBSD, however, issetugid() returns non-zero even
+     * if the process started as root, not-set-uid, and then later
+     * called seteuid(), for example, but in that case we'd want to
+     * trust the environ!  So if issetugid() > 0 we want to do something
+     * else.  See below.
      */
-    return issetugid();
+    if (issetugid() == 0)
+        return we_are_suid = 0;
 #endif /* USE_RK_GETAUXVAL */
 
+#if defined(AT_EXECFN) || defined(AT_EXECPATH)
+
+  /*
+   * There's an auxval by which to find the path of the program this
+   * process exec'ed.
+   *
+   * Linux calls this AT_EXECFN.  FreeBSD calls it AT_EXECPATH.  NetBSD
+   * and Illumos call it AT_SUN_EXECNAME.
+   *
+   * We can stat it.  If the program did a chroot() and the chroot has
+   * a program with the same path but not set-uid/set-gid, of course,
+   * we lose here.  But a) that's a bit of a stretch, b) there's not
+   * much more we can do here.
+   */
+#if defined(AT_EXECFN) && !defined(AT_EXECPATH)
+#define AT_EXECPATH AT_EXECFN
+#endif
+#if defined(AT_SUN_EXECNAME) && !defined(AT_EXECPATH)
+#define AT_EXECPATH AT_EXECFN
+#endif
+    {
+        unsigned long p = getauxval(AT_EXECPATH);
+        struct stat st;
+        
+        if (p != 0 && *(const char *)p == '/' &&
+            stat((const char *)p, &st) == 0) {
+            if ((st.st_mode & S_ISUID)) {
+                errno = save_errno;
+                return we_are_suid = 1;
+            }
+            if ((st.st_mode & S_ISGID)) {
+                errno = save_errno;
+                return we_are_suid = 2;
+            }
+            errno = save_errno;
+            return we_are_suid = 0;
+        }
+    }
+#endif
+
+    /*
+     * Fall through if we have rk_getauxval() but we didn't have (or
+     * don't know if we don't have) the aux entries that we needed.
+     * We're done with it.
+     */
+
+#if defined(HAVE_ISSETUGID)
+    if (issetugid()) {
```

Comment:
Yes.  I can return unconditionally here, but only if I add a `#else` to avoid dead code warnings... though... if the compiler is smart enough it would emit them anyways, so, ok, I'll add the `#else`.